File: modules/preprocessing/pipelines/resamplingdemo.lua
require("cgt")
require("base")
require("cvio")
require("vis")
require("preprocessing")

pipeline = campvis.newPipeline("ResamplingDemoLua")

function pipeline:ctor()
    print("I'm being constructed!")
    self.lsp = base.LightSourceProvider()
    self.addProcessor(instance, self.lsp)

    self.image_reader = cvio.MhdImageReader()
    self.addProcessor(instance, self.image_reader)

    local canvas_size = self.getProperty(instance, "CanvasSize")
    self.ve = vis.VolumeExplorer(canvas_size)
    self.resampler = preprocessing.GlImageResampler(canvas_size)

    self.addProcessor(instance, self.resampler)
    self.addProcessor(instance, self.ve)

    self.addEventListenerToBack(instance, self.ve)
end

function pipeline:init()
    print("I'm being inited!")

    self.ve.p_outputImage:setValue("result")
    self.getProperty(instance, "renderTargetID"):setValue("result")

    self.image_reader.p_url:setValue(campvis.SOURCE_DIR .. "/modules/vis/sampledata/smallHeart.mhd")
    self.image_reader.p_targetImageID:setValue("reader.output")
    self.image_reader.p_targetImageID:addSharedProperty(self.resampler.p_inputImage)

    self.resampler.p_outputImage:setValue("resampled")
    self.resampler.p_outputImage:addSharedProperty(self.ve.p_inputVolume)

    local geometry1 = campvis.TFGeometry1D_createQuad(cgt.vec2(0.1, 0.125), cgt.col4(255, 0, 0, 32),
                                                      cgt.vec4(255, 0, 0, 32))
    local geometry2 = campvis.TFGeometry1D_createQuad(cgt.vec2(.4, .5), cgt.col4(0, 255, 0, 128),
                                                      cgt.vec4(0, 255, 0, 128))

    local dvrTF = campvis.Geometry1DTransferFunction(128, cgt.vec2(0, 0.05))
    dvrTF:addGeometry(geometry1)
    dvrTF:addGeometry(geometry2)

    self.ve:getNestedProperty("VolumeRendererProperties::RaycasterProps::TransferFunction"):replaceTF(dvrTF)

    self.resampler.p_resampleScale:setValue(1.25)
end

function pipeline:deinit()
    print("I'm being deinited!")
end

File: modules/preprocessing/pipelines/resamplingdemo.cpp
// ================================================================================================
// 
// This file is part of the CAMPVis Software Framework.
// 
// If not explicitly stated otherwise: Copyright (C) 2012-2014, all rights reserved,
//      Christian Schulte zu Berge <christian.szb@in.tum.de>
//      Chair for Computer Aided Medical Procedures
//      Technische Universitaet Muenchen
//      Boltzmannstr. 3, 85748 Garching b. Muenchen, Germany
// 
// For a full list of authors and contributors, please refer to the file "AUTHORS.txt".
// 
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file 
// except in compliance with the License. You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software distributed under the 
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, 
// either express or implied. See the License for the specific language governing permissions 
// and limitations under the License.
// 
// ================================================================================================

#include "resamplingdemo.h"

#include "tgt/event/keyevent.h"
#include "core/datastructures/imagedata.h"
#include "core/datastructures/renderdata.h"

#include "core/classification/geometry1dtransferfunction.h"
#include "core/classification/tfgeometry1d.h"
#include "core/tools/glreduction.h"


namespace campvis {

    ResamplingDemo::ResamplingDemo(DataContainer* dc)
        : AutoEvaluationPipeline(dc)
        , _lsp()
        , _imageReader()
        , _resampler(&_canvasSize)
        , _ve(&_canvasSize)
    {
        addProcessor(&_lsp);
        addProcessor(&_imageReader);
        addProcessor(&_resampler);
        addProcessor(&_ve);

        addEventListenerToBack(&_ve);
    }

    ResamplingDemo::~ResamplingDemo() {
    }

    void ResamplingDemo::init() {
        AutoEvaluationPipeline::init();

        _ve.p_outputImage.setValue("result");
        _renderTargetID.setValue("result");

        _imageReader.p_url.setValue(CAMPVIS_SOURCE_DIR "/modules/vis/sampledata/smallHeart.mhd");
        _imageReader.p_targetImageID.setValue("reader.output");
        _imageReader.p_targetImageID.addSharedProperty(&_resampler.p_inputImage);

        _resampler.p_outputImage.setValue("resampled");
        _resampler.p_outputImage.addSharedProperty(&_ve.p_inputVolume);
    
        Geometry1DTransferFunction* dvrTF = new Geometry1DTransferFunction(128, tgt::vec2(0.f, .05f));
        dvrTF->addGeometry(TFGeometry1D::createQuad(tgt::vec2(.1f, .125f), tgt::col4(255, 0, 0, 32), tgt::col4(255, 0, 0, 32)));
        dvrTF->addGeometry(TFGeometry1D::createQuad(tgt::vec2(.4f, .5f), tgt::col4(0, 255, 0, 128), tgt::col4(0, 255, 0, 128)));
        static_cast<TransferFunctionProperty*>(_ve.getNestedProperty("VolumeRendererProperties::RaycasterProps::TransferFunction"))->replaceTF(dvrTF);
    }

    void ResamplingDemo::deinit() {
        _canvasSize.s_changed.disconnect(this);
        AutoEvaluationPipeline::deinit();
    }


}